User command: parsing logic + schema creation

muzarski opened the pull request:
ref: https://cassandra.apache.org/doc/stable/cassandra/tools/cassandra_stress.html#user-mode

## Motivation

User command is the most complex c-s command. It is very flexible and allows user to define his own schema, as well as stress queries. We obviously want to support this command in cql-stress.

### User profiles
ref: https://cassandra.apache.org/doc/stable/cassandra/tools/cassandra_stress.html#profile

User can define the benchmark via profile yaml file. In this PR we introduce support for most commonly used profile parameters, which are:
- `keyspace`
- `keyspace_definition`
- `table`
- `table_definition`
- `queries`

## Changes 
- Introduced `UserParams` struct which contains the payload needed to run the benchmark
- Implemented CLI parsing logic for `user` command, which includes the `profile=` parameter that specifies the path to yaml file. Notice that c-s supports `ops` and `clustering=` parameters as well. These two define the ratio which we sample the specific `queries` from. As for now, I want to keep it simple and in the following PRs I plan to sample the operations in the round-robin fashion. We will introduce non-determinism in the future.
- Implemented the very basic runtime logic which prepares the schema for benchmark

## Tests
I think I could add some tests with exemplary yaml files. I'm not sure if it's a good idea since this would pollute our repo a bit (?). WDYT?

wprzytula commented:
> I think I could add some tests with exemplary yaml files. I'm not sure if it's a good idea since this would pollute our repo a bit (?). WDYT?

Definitely, do that! Those will serve both as test cases and for quick reference on what kind of user config is supported.

muzarski commented on Cargo.toml:
```diff
@@ -35,6 +35,8 @@ tokio = { version = "1.15.0", features = [
 tracing = "0.1.35"
 tracing-subscriber = { version = "0.3.17", features = ["env-filter"] }
 rust-strictmath = "0.1.1"
+serde = { version = "1.0", features = ["derive"] }
+serde_yaml = "0.8"
```
User profiles are broadly used in SCT as this is the only mode that allows to define custom schemas. Simple modes, such as `write/read` always create table with only `blob` columns.

OTOH, some users may only want to perform simple benchmarks (for these, `write/read` workloads are enough) - it would make sense to hide user profiles behind a feature then.

muzarski commented on src/bin/cql-stress-cassandra-stress/settings/command/user.rs:
```diff
@@ -18,3 +21,12 @@ pub struct QueryDefinitionYaml {
+
+    fn parse(s: &str) -> Result<Self::Parsed> {
+        let yaml = File::open(s).with_context(|| format!("Invalid profile yaml file: {}", s))?;
+        serde_yaml::from_reader(yaml).context("Failed to parse profile yaml file")
```
Yes, the error message passed to `context()` is pushed on top of the stack. When program terminates with the error, whole stack is displayed.

muzarski commented on src/bin/cql-stress-cassandra-stress/settings/command/mod.rs:
```diff
@@ -71,6 +73,7 @@ impl Command {
             Command::CounterWrite => "Multiple concurrent updates of counters.",
             Command::CounterRead => "Multiple concurrent reads of counters. The cluster must first be populated by a counterwrite test.",
             Command::Mixed => "Interleaving of any basic commands, with configurable ratio and distribution - the cluster must first be populated by a write test.",
+            Command::User => "Interleaving of user provided queries, with configurable ratio and distribution",
```
As of now, there is no need to require that.

In c-s, `user` mode has no validation logic for the inserted data - this means that populating the cluster is not necessary.

However, looking at c-s code, it seems that someone started implementing data validation. Unfortunately, it seems to fail during parsing, and obviously, it's not documented. There is an idea to add such feature to cql-stress, though. Do you think it's better to add such requirement right now, even though it will only be required in the future?

wprzytula replied:
In general, yes; let's introduce requirements early, so that developers never have to change their habits and/or code once they get used to them.

muzarski commented on src/bin/cql-stress-cassandra-stress/settings/command/user.rs:
```diff
@@ -67,3 +73,73 @@ pub struct UserParams {
+                    Err(_) => anyhow::bail!("Failed to parse query definition"),
+                }
+            })
+            .collect::<Result<BTreeMap<_, _>, _>>()?;
```
I addressed above comment and got rid of `anyhow::bail!` here. I'll mark is as resolved.

muzarski commented on src/bin/cql-stress-cassandra-stress/settings/command/user.rs:
```diff
@@ -67,3 +73,73 @@ pub struct UserParams {
+    for group in groups.iter_mut() {
+        group.push(Box::new(profile.clone()));
+        parser.group(&group.iter().map(|e| e.as_ref()).collect::<Vec<_>>());
+    }
```
Thank you for the suggestion!

muzarski commented on src/bin/cql-stress-cassandra-stress/settings/command/user.rs:
```diff
@@ -67,3 +73,73 @@ pub struct UserParams {
+    }
+
+    fn parse_with_handles(handles: UserParamHandles) -> Result<Self> {
+        let profile = handles.profile.get().unwrap();
```
Honestly, it is the first time I see something like that. Thank you for sharing it with me.

muzarski commented:
v2:
- rebased on top of master (unfortunately, it is contained in the same diff as other changes - I apologize for that)
- addressed review comments
- hidden user profiles logic behind a crate feature
- adjusted CI, so we build the tool with and without the feature

wprzytula commented:
Is the list of features documented somewhere?
If we support something (in this case: user profiles) only conditionally, it should definitely be said that you can turn it using _name-of-the-feature_.

muzarski commented:
v2.1: moved `queries.is_empty()` check, so it is detected during yaml file parsing.

muzarski commented:
> > I think I could add some tests with exemplary yaml files. I'm not sure if it's a good idea since this would pollute our repo a bit (?). WDYT?
> 
> Definitely, do that! Those will serve both as test cases and for quick reference on what kind of user config is supported.

v3: added test cases for contents of yaml files

muzarski commented:
v3.1: adjusted the logic so unknown yaml parameters are denied and treated as error.

muzarski commented:
> Is the list of features documented somewhere? If we support something (in this case: user profiles) only conditionally, it should definitely be said that you can turn it using _name-of-the-feature_.

The problem is that `README` for c-s frontend is just empty. I'll open a PR that updates the README (and includes info about user profiles and set of supported features) which will base on this PR. WDYT?
